src/compiler/transformers/declarations.ts

```diff
@@ -254,11 +257,11 @@ export function transformDeclarations(context: TransformationContext) {
     let needsScopeFixMarker = false;
     let resultHasScopeMarker = false;
     let enclosingDeclaration: Node;
-    let necessaryTypeReferences: Set<[specifier: string, mode: ResolutionMode]> | undefined;
+    let necessaryTypeReferences: Map<ModeAwareCacheKey, [specifier: string, mode: ResolutionMode]> | undefined;

```

Comment:
Since we keep all references as they are declared in source, we can't use the tuple as the key because:

1. We mark all source references as necessary upfront.
2. Even if the host can't resolve the reference, we should still preserve it to keep as close to the original source.

An alternate approach would be to ask the host for the `[specifier, mode]` tuple. This approach however fells error prone, since even in current emit some of these tuples come from the host, while other are created in the declaration transform.